packages/database/src/migrations/20210816230203-MoveAnalyticsTableUtilityFunctionsToDatabaseFunctions-modifies-schema.js

```diff
@@ -0,0 +1,250 @@
+          to_char(survey_response.data_time, ''YYYYMM'') as "month_period",
+          to_char(survey_response.data_time, ''YYYY'') as "year_period",
+          survey_response.data_time as "date"
+        FROM
+          survey_response
+        INNER JOIN
+          answer ON answer.survey_response_id = survey_response.id
+        INNER JOIN
+          entity ON entity.id = survey_response.entity_id
+        INNER JOIN
+          survey ON survey.id = survey_response.survey_id
+        INNER JOIN
+          question ON question.id = answer.question_id
+        INNER JOIN
+          data_source ON data_source.id = question.data_source_id
+        WHERE data_source.service_type = ''tupaia'' AND survey_response.outdated IS FALSE';
+    
+    begin
+      RAISE NOTICE 'Creating Materialized View Logs...';
+    
+      FOREACH source_table IN ARRAY source_tables_array LOOP
+        IF (SELECT NOT EXISTS (
+          SELECT FROM pg_tables
+          WHERE schemaname = 'public'
+          AND tablename   = 'log$_' || source_table
+        ))
+        THEN
+          EXECUTE 'ALTER TABLE ' || source_table || ' DISABLE TRIGGER ' || source_table || '_trigger';
+          tStartTime := clock_timestamp();
+          PERFORM mv$createMaterializedViewlog(source_table, 'public');
+          RAISE NOTICE 'Created Materialized View Log for % table, took %', source_table, clock_timestamp() - tStartTime;
+          EXECUTE 'ALTER TABLE ' || source_table || ' ENABLE TRIGGER ' || source_table || '_trigger';
+        ELSE
+          RAISE NOTICE 'Materialized View Log for % table already exists, skipping', source_table;
+        END IF;
+      END LOOP;
+    
+    
+      RAISE NOTICE 'Creating analytics materialized view...';
+      IF (SELECT NOT EXISTS (
+        SELECT FROM pg_tables
+        WHERE schemaname = 'public'
+        AND tablename   = 'analytics'
+      ))
+      THEN
+        tStartTime := clock_timestamp();
+        PERFORM mv$createMaterializedView(
+            pViewName           => 'analytics',
+            pSelectStatement    =>  pSqlStatement,
+            pOwner              => 'public',
+            pFastRefresh        =>  TRUE
+        );
+        RAISE NOTICE 'Created analytics table, took %', clock_timestamp() - tStartTime;
+      ELSE
+        IF (force)
+        THEN
+        RAISE NOTICE 'Force rebuilding analytics table';
+          tStartTime := clock_timestamp();
+          PERFORM mv$createMaterializedView(
+              pViewName           => 'analytics_tmp',
+              pSelectStatement    =>  pSqlStatement,
+              pOwner              => 'public',
+              pFastRefresh        =>  TRUE
+          );
+          RAISE NOTICE 'Created analytics table, took %', clock_timestamp() - tStartTime;
+          PERFORM mv$removeMaterializedView('analytics', 'public');
+          PERFORM mv$renameMaterializedView('analytics_tmp', 'analytics', 'public');
```

Comment:
This is using a new function in `pg-mv-fast-refresh` that has been added to rename a MV. Seen here: https://github.com/beyondessential/pg-mv-fast-refresh/pull/4